rfcs/0000-structured-logging.md

```diff
@@ -0,0 +1,2189 @@
+    message = "message";
+    correlation = correlation_id,
+    user
+);
+```
+
+Any `value` or `key = value` expressions before the `;` in the macro will be interpolated into the message as unstructured text. This is the `log!` macro we have today. Any `value` or `key = value` expressions after the `;` will be captured as structured key-value pairs. These structured key-value pairs can be inspected or serialized, retaining some notion of their original type. That means in the above example, the `message` key is unstructured, and the `correlation` and `user` keys are structured:
+
+```
+info!(
+    "This is the rendered {message}. It is not structured",
+    message = "message";
+    ^^^^^^^^^^^^^^^^^^^
+    unstructured
+
+    correlation = correlation_id,
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    structured
+
+    user
+    ^^^^
+    structured
+);
+```
+
+### What can be logged?
+
+Anything that's `std::fmt::Debug + serde::Serialize` can be logged as a structured value. Using default crate features though, only a fixed set of types from the standard library are supported:
+
+- Standard formats: `Arguments`
+- Primitives: `bool`, `char`
+- Unsigned integers: `u8`, `u16`, `u32`, `u64`, `u128`
+- Signed integers: `i8`, `i16`, `i32`, `i64`, `i128`
+- Strings: `&str`, `String`
+- Bytes: `&[u8]`, `Vec<u8>`
+- Paths: `Path`, `PathBuf`
+- Special types: `Option<T>` and `()`
+
+This keeps the `log` crate lightweight by default. In the example from before, `correlation_id` and `user` can be used as structured values if they're in that set of types:
+
+```rust
+let user = "a user id";
+let correlation_id = "some correlation id";
+
+info!(
+    "This is the rendered {message}. It is not structured",
+    message = "message";
+    correlation = correlation_id,
+    user
+);
+```
+
+What if the `correlation_id` is a `uuid::Uuid` instead of a string? What if the `user` is some other datastructure containing an id along with some other metadata? We still want to be able to log these values. The set of loggable values above can be expanded to also include any other type that implements both `std::fmt::Debug` and `serde::Serialize` using a crate feature:
+
+```toml
+[dependencies.log]
+features = ["kv_serde"]
+
+[dependencies.uuid]
+features = ["serde"]
+```
+
+```rust
+#[derive(Debug, Serialize)]
+struct User {
+    id: Uuid,
+    ..
+}
+
+let user = User { id, .. };
+let correlation_id = Uuid::new_v4();
+
+info!(
+    "This is the rendered {message}. It is not structured",
+    message = "message";
+    correlation = correlation_id,
+    user
+);
+```
+
+So the effective trait bounds for structured values are `Debug + Serialize`:
+
+```
+info!(
+    "This is the rendered {message}. It is not structured",
+    message = "message";
+              ^^^^^^^^^
+              Display
+
+    correlation = correlation_id,
+                  ^^^^^^^^^^^^^^
+                  Debug + Serialize
+
+    user
+    ^^^^
+    Debug + Serialize
+);
+```
```

Comment:
How does this look on the loggers side? Is `user` flattened? Or does this create a key "user" -> value `user` structure mapping?

Reply:
If the `kv_serde` feature is enabled then `user` will be serialized as a map, not flattened. If the `kv_serde` feature is not enabled then `user` will be serialized as a string using its `Debug` implementation.